### ISSUE TITLE
Add support for VSS (windows)

### ISSUE THREAD

**Issue by frederich** [restic/restic#340]
The restic backup runs. In the mean time my work is going on. Restic come to the point where it saves my Outlook archive. But Outlook is still open, restic fails.

```
[2Kscanned 77389 directories, 924469 files in 7:49 GiBerror for e:\home\jf\archive\outlook\archive.pst: SaveFile() chunker.Next(): read e:\home\jf\archive\outlook\archive.pst: Der Prozess kann nicht auf die Datei zugreifen, da ein anderer Prozess e
```

**Comment by fd0** [comment]
Hm, what do other programs on Windows typically do? It looks like restic is denied opening the file...

**Comment by episource** [comment]
The way to go is Volume Shadow Copy Service (VSS) [1, 2]. It creates (at least) crash consistent copy-on-write snapshots and allows safely copying opened files. Additionally The Volume Shadow Copy service notifies compatible applications (like MS SQL Server) that a VSS Snapshot is to be taken so that they can provide consistent data to be included in the snapshot.

VSS support could be integrated into restic, but it's also possible to create a VSS snapshot externally and use that as a backup source. [3,4]  might be of interest.

[1] https://technet.microsoft.com/en-us/library/ee923636.aspx
[2] https://en.wikipedia.org/wiki/Shadow_Copy
[3] https://technet.microsoft.com/en-us/library/cc754968.aspx
[4] http://www.heise.de/ct/inhalt/2009/09/180/

**Comment by fd0** [comment]
Thanks for linking the documentation! This looks like a big endeavour, though.

**Comment by episource** [comment]
If someone is willing to integrate VSS directly into restic, the VSS API description can be found [here](https://msdn.microsoft.com/en-us/library/aa384648.aspx).

**Comment by turnkey-commerce** [comment]
To more easily integrate restic with automatic VSS creation I wrote a Windows PowerShell script that can be run to automate this without the complications mentioned above.  It is available in the attached zip: 
[restic-backup-windows-vss.zip](https://github.com/restic/restic/files/1066538/restic-backup-windows-vss.zip)

If restic has a repository for external scripts but perhaps it could be included there.

The meat of the script is in these lines excerpted from the script (omitting the config part):

```
$ShadowPath = $rootVolume + 'shadowcopy\'

# Create a volume shadow copy and make it accessible
$s1 = (Get-WmiObject -List Win32_ShadowCopy).Create($rootVolume, "ClientAccessible")
$s2 = Get-WmiObject Win32_ShadowCopy | Where-Object { $_.ID -eq $s1.ShadowID }
$device  = $s2.DeviceObject + "\"

# Create a symbolic link to the shadow copy
cmd /c mklink /d $ShadowPath "$device"

# Run Restic on the data files in the shadowcopy

ForEach ($folderToBackup in $foldersToBackup) {
    cmd /c $resticExe backup -r $resticRepository ($ShadowPath + $folderToBackup)
}

# Delete the shadow copy and remove the symbolic link
$s2.Delete()
cmd /c rmdir $ShadowPath
```

**Comment by fd0** [comment]
Thank you very much! There's no `contrib` repo of some kind, but I'm sure that interested users will find this issue and your script.

**Comment by turnkey-commerce** [comment]
You're welcome and thanks for sharing this useful application!

**Comment by bwmarrin** [comment]
That's a nice script @turnkey-commerce  :) I've been using normal non-powershell batch file to do this with rsync, but your version is nicer.

On the topic of restic though.  Adding VSS support somehow would be really really OMG really nice :)  Without that, it's hard to break into the Windows Enterprise segment of backups where VSS is both expected and really needed.  Of course, shell script work arounds like above will help mitigate that, especially if decently documented somewhere :)

**Comment by jcasale** [comment]
Keep in mind when implementing this feature, if you are attempting to access a file in a shadow copy whose provider has data on other volumes in addition to the one you snapshot, the writer will be excluded from the shadow copy unless you add each volume that contains data.

The WMI example won't suffice as the Create signature only facilitates a single volume.

**Comment by aventrax** [comment]
The script proposed by @turnkey-commerce work great. I have been able to schedule it on machine startup to backup some stuff including locked files. Anyway, I think that VSS must be used by a user with admin privileges (in elevated mode) to work, so restic should be able to manage this as "sudo" does on linux.

Moreover, It's true that the VSS support and a small set of remote commands for clients and repository (API), would be necessary to implement this in a company.

@fd0 : Great job with Restic, I discovered it a few weeks ago and it works great!

**Comment by bjoe2k4** [comment]
I am trying to use the Powershell script by @turnkey-commerce, which works great, but i am having trouble writing the restic output to a logfile. No matter what i do, i always end up having a jammed logfile. For example, when replacing line 33 with:
```
cmd /c $resticExe backup -r $resticRepository ($ShadowPath + $folderToBackup) >> backup.log
```
there goes something horribly wrong:
```

[2K[0:01] 0 files 0 B, total 1 files 11 B, 0 errors

[2K[0:01] 15 files 199 B, total 25 files 995.782 KiB, 0 errors

[2K
[2K[0:01] 15 files 199 B, total 25 files 995.782 KiB, 0 errors
... and so on...
```
However, when i run a bare restic backup command on cmd.exe and redirect that to a logfile (here having `-v -v` options) all is fine.
```
open repository
lock repository
load index files
using parent snapshot 2709b9ff
start scan on [C:\ProgramData]
start backup on [C:\ProgramData]
unchanged /C/ProgramData/.asv2sfi
unchanged /C/ProgramData/.cld2sfi
unchanged /C/ProgramData/.dys1sfi
... and so on...
```

**Comment by fd0** [comment]
Good point, thanks! I'm also not sure it's a good idea to add this to restic, so I'll move this to the "feature request" stage.

**Comment by mgumz** [comment]
Well, it is more a "improve the documentation" thing than a feature. Especially if you think about the use cases and the given different filesystems which are able to create a snapshot. Tomorrow someone appears and wants snapshots of ReFS on Windows, UFS2 snapshots on FreeBSD, APFS from Apple etc :)

**Comment by bjoe2k4** [comment]
> @bjoe2k4 try the cmd /c ... part without all the rest and see if your "logging problems" persists.

The problems persist. I agree that this has nothing to do with VSS, I rather didn't want to open a new issue for this minor problem, as I was expecting some error on my side.

In the meantime I garthered new insights and also a workaround regarding my logging problems. It seems to be a interop problem with powershell and restic. I will open a new issue when i find time.

I also do agree that "managing snapshots of a filesystem" is outside the scope of restic.

**Comment by matthijskooijman** [comment]
I wonder if this isn't something to put into restic after all. I can see that restic should not be trying to automatically manage snapshots on all kinds of systems. However:
 - I believe that Windows is a lot more picky about reading opened files, making it more important to use snapshotting to get a complete backup (so it's about completeness, not just consistency).
 - VSS is, AFAICS, a generic FS-agnostic feature, that handles this issue completely for the Windows platform, which makes it an implementation with a high bang-for-buck.
 - VSS is, AFAICS from the powershell example, fairly straightforward to get to work (unlike snapshotting on unix system snapshot which AFAIK are either filesystem-specific or operate on the block level, requiring a lot more insight in how the system is set up).

Additionally, doing this natively in restic seems a lot more reliable than from a script (in particular in terms of proper cleanup and guaranteeing a snapshot is removed after usage).

**Comment by matthijskooijman** [comment]
One more consideration I found when looking more into VSS, is the handling or cleanup. Ideally, you want a non-persistent shadow copy, meaning it gets automatically cleaned up when it is released, which seems like an important feature to prevent shadow copies from piling up and taking up disk space. I assume this also means they are removed when the backup process somehow crashes or otherwise terminates without explicitly cleaning up, though I could not find definitive info on that (though non-persistent shadow copies *are* always cleaned up on reboot). See auto-release shadow copy on https://docs.microsoft.com/en-us/windows/desktop/vss/vssgloss-ahttps://docs.microsoft.com/en-us/windows/desktop/vss/vssgloss-a and all types on https://docs.microsoft.com/en-us/windows/desktop/api/Vss/ne-vss-_vss_snapshot_context

The powershell script shown above uses a "ClientAccessible" context. I'm not really sure what that means exactly, but it results in a persistent snapshot according to `vssadmin list shadows` (and I also confirmed that killing restic with ^C leaves the shadow copy behind). I tried modifying it to "Backup" (based on [this doc](https://docs.microsoft.com/en-us/windows/desktop/api/Vss/ne-vss-_vss_snapshot_context)) or "Volatile" (based on [this doc](https://docs.microsoft.com/en-us/windows-server/administration/windows-commands/set-context#parameters)), but both of these return code 5, which means ["Unsupported shadow copy context"](https://docs.microsoft.com/en-us/previous-versions/windows/desktop/vsswmi/create-method-in-class-win32-shadowcopy#return-value). I'm not sure if I'm just using the wrong vocabulary, or non-persistent types are not supported. [This post](https://social.technet.microsoft.com/Forums/ie/en-US/4de900bc-95fa-499b-b25a-1ec787efa1e1/win32shadowcopy-unsupported-shadow-copy-context?forum=winserverpowershell#d75aa7fa-5d1e-4992-93f0-c9b6eea1e72c) suggests that this type might not be supported for workstations, but I somehow doubt that (also because [these docs](https://docs.microsoft.com/en-us/windows/desktop/api/Vss/ne-vss-_vss_snapshot_context#remarks) say that `VSS_CTX_BACKUP` is the default, and also the only supported value on Windows XP).

A related observation is that you might [need a persistent shadow copy](https://docs.microsoft.com/en-us/windows/desktop/vss/vshadow-tool-and-sample#exposing-a-shadow-copy-locally) to expose a shadow copy on a drive letter or folder. I somewhat suspect that the `mklink` approach used by the powershell script might also work on non-persistent snapshots. If it does not, that might be another reason to support VSS in restic - so restic can access the shadow copy directly, without needing any drive letter or link (though I think that accessing VSS files is a matter of using a magic filename, like `\\?\GLOBALROOT\Device\HarddiskVolumeShadowCopy123`, but that needs some kind of path-mapping like #1376 suggests)

Note even with the link approach in the above Powershell script there is still a caveat: When trying to backup the entire volume, the path to the link is passed to restic, which does not follow it (which relates to #542).

**Comment by cfbao** [comment]
> When trying to backup the entire volume, the path to the link is passed to restic, which does not follow it

You can modify the script to first `cd` into the symlink, then use `backup .` to backup the entire shadow copy.

**Comment by jarmo** [comment]
Just as an information [Duplicati](https://github.com/duplicati/duplicati/blob/0a783a6e2faf37212e8a59d3018637170e987dd2/Duplicati/Library/Snapshots/WindowsSnapshot.cs) has VSS support built into it, which is really nice. However, it has multiple other problems why I've been looking for a different solution and restic seems to be great after some testing! Adding VSS into it would make it (maybe?) perfect.

**Comment by jarmo** [comment]
I changed @turnkey-commerce script to this:

```powershell
$disk = $args[0]
$shadowFilePath = $args[1]
$s1 = (Get-WmiObject -List Win32_ShadowCopy).Create($disk, "ClientAccessible")
$s2 = Get-WmiObject Win32_ShadowCopy | Where-Object { $_.ID -eq $s1.ShadowID }

@{
  id = $s2.ID
  deviceObject = $s2.DeviceObject
} | ConvertTo-Json | Out-File -Encoding "Default" $shadowFilePath
```

And I'm calling it like this:

```
powershell.exe -ExecutionPolicy Bypass -File create-vss.ps1 c:\ c-shadow.json
```

This allows my backup script to cleanup VSS shadow copy before and after making a new one, which solves the problem of ^C or any other interrupted exits.

**Comment by bherila** [comment]
VSS is something so commonly needed while backing up on Windows that I agree it makes sense to include in restic.

**Comment by jarmo** [comment]
Actually, after further thinking I think that this is rather good thing that VSS is not included. This keeps backup software itself less complex and gives you all the necessary tools to use it if you need. Managing it with Powershell is really easy. However, maybe this should be somewhere documented so that it would be easier to notice.

**Comment by rawtaz** [comment]
On the other hand, having to learn Powershell just to back up properly on Windows is a pretty big no-no. It might be easy for you, but for many others Powershell is horribly convoluted and hard to get your head around.

**Comment by bherila** [comment]
Also, if you're *not* calling restic from PowerShell (what about another program or script? What about bash on Windows?) or some environment where the VSS APIs are complex and error handling can also add an additional layer of mess. I think VSS support is critical to making restic "just work" on Windows.

**Comment by jarmo** [comment]
@rawtaz @bherila that's why I did create as minimal Powershell script as possible and it only creates VSS volume. This Powershell script and everything else including restic is called from a Ruby script (in my particular case, but it can whatever scripting language you prefer). For cleaning up VSS volume, I parse that JSON (you can even use jq in bash, if you don't like to use any other scripting languages ;)) and call `vssadmin delete shadows` before and after backup. That's it. Powershell script is only needed to create VSS shadow since it would be much more complicated to do directly in Ruby.

That's pretty easy while keeping Restic as much of a file-system agnostic as possible.

**Comment by bherila** [comment]
Having to write a script and parse JSON sounds like a lot of work for a restic user to take on. VSS is a Windows specific feature - not a file system specific feature. Since VSS is virtually required to create a reasonable backup of any Windows machine (even a user's home "profile" directory), I think the use case is pretty strong for having VSS support built in.

Looking at this another way, restic already builds Windows binaries which are specific to Windows, do you really think it is reasonable to say "Ok we are providing this binary for Windows but you can't really use it unless you are backing up an offline machine or you write a script to wrap the VSS provider." In that case, restic.exe is not a standalone utility but merely a backup library! And I really, really believe this severely limits its usefulness and its audience on Windows.

**Comment by mholt** [comment]
I would like what articles you recommend for learning about VSS -- why is it necessary? Does Windows not allow reading files that are open at all by another application? What are the recommended reading for this (beyond just googling it, obviously)?

**Comment by bherila** [comment]
The purpose of VSS is to provide a systemwide mechanism for applications to flush pending transactions to disk in a consistent state. For example, if you had a database transaction in progress, and you capture the database file mid-transaction, the file might be backed up in a corrupted state. 

With VSS, an application can be made aware of requests by the OS to read its files. Then, the OS can notify the VSS-aware application "I want to read your file" and then the OS can say "ok, here's a snapshot of this file that's safe to back up." 

Virtually all enterprise-grade backup solutions for Windows implement VSS support.

Further reading: 
* https://docs.microsoft.com/en-us/windows/desktop/VSS/common-volume-backup-issues

**Comment by cfbao** [comment]
I think a practical question is: _how_ can restic implement VSS?

Duplicati uses C#, so it should be quite straightforward for them to implement VSS.
But restic is written in Go, and tries to remain pure Go to facilitate cross compilation and distribution. I've not seen any documentation on implementing VSS in a portable language like Go. Also main developer @fd0 doesn't use Windows, whereas VSS is quite deep in the Windows water. These factors make implementing VSS _very_ difficult.

Any suggestions?

**Comment by matthijskooijman** [comment]
> Does Windows not allow reading files that are open at all by another application?

This is how I understand it, yes,

**Comment by bjoe2k4** [comment]
One also might just use the `VShadow` tools form the Windows SDK for creating the shadowcopy. [Link](https://docs.microsoft.com/en-us/windows/desktop/vss/vshadow-tool-and-sample)

Should be pretty straightforward to call from any language and is for example used in `c't WIMage`, which is a backup solution developed by a german computer magazine.

**Comment by mholt** [comment]
Yeah, if there's a standard command that comes with Windows, restic could invoke that externally, potentially. That might be preferable to trying to re-implement it in Go.

**Comment by cfbao** [comment]
@bjoe2k4 @mholt The only viable option I can find is still PowerShell, which is probably why if you Google "creating shadow copy", the most common answers all use PowerShell.
All other tools I've seen have issues that make them unsuitable:
* _VShadow_ is not built-in. Asking users to install an SDK just for VSS doesn't sound like a good idea. Also it's version specific, and quite easy for users to install a wrong/unsupported version.
* _VssAdmin_ is built-in, but only supports shadow copy _creation_ on Windows Server. On other versions (Win 7/8/10 etc.) it's more of a inquiry tool
* _DiskShadow_ is only found on Windows Server.

----
> Does Windows not allow reading files that are open at all by another application?

@mholt @matthijskooijman Most opened files can be read just fine. Only if the opening application explicitly lock the file will reading become impossible. I typically find locked files in the TEMP folder, and actually haven't encountered a case when I need to backup a locked file. But I don't use Outlook (which prompted this issue originally).
I find VSS more useful at backing up live databases (which is actually _not_ system locked) to prevent corruption.

**Comment by bherila** [comment]
The lowest level abstraction for VSS is WMI, which is controlled via COM. Unfortunately, Go doesn't have any built in wrappers for COM, like C# has. Without knowing much about Go myself, it might be possible to use https://github.com/go-ole/go-ole or https://github.com/StackExchange/wmi/blob/master/wmi.go but we are approaching the edge of my knowledge at this point!

**Comment by fd0** [comment]
Another issue that makes me wonder if it is a good idea to have VSS integrated into restic is: how would users configure it? From what you're describing it sounds like there's a high chance that we will run into a configuration nightmare...

**Comment by matthijskooijman** [comment]
> how would users configure it?

I *believe* that no real configuration is needed. I think Restic could:
 - Figure out what volumes are involved in the backup
 - Create a shadow copy of the backup
 - Backup files from the shadow copy (translating pathnames internally, so there is no need to symlink or mount the shadow copy anywhere)
 - Remove the shadow copy (ideally, it should be autoremoved, but I'm not sure that's always possible, see https://github.com/restic/restic/issues/340#issuecomment-417795104)

I suspect this is also what most other backup solutions do, though I haven't got much experience (I'm not really a Windows user, though I do have an interest in backup someone else's Windows laptop).

**Comment by cfbao** [comment]
I agree with @matthijskooijman that no user configuration is required, and his list of steps is also how I imagine restic would work with VSS.

The complications mentioned earlier is more about implementation, in particular the various Windows tools restic may or may not use. Once it's implemented, it should be just one more flag for Windows users to specify if they want to use VSS.

restic only needs to report appropriate errors when things go south, e.g. lack of privilege, volume not eligible for VSS... It's usually expected that such errors are handled outside of the backup program.

**Comment by jarmo** [comment]
I totally understand everyone's painpoints of Restic not having VSS support built-in and I have to admit that this was also a con for myself aswell before starting using Restic. However, I've used Duplicati for years now and I find one of its biggest problems to be that it supports _everything_. This has had a negative impact to Duplicati since it has gotten really complex over the years - complexity affects its codebase and its UI/GUI too. This means that every new version update might break something. I've had to wait weeks before getting anything restored and I just don't want to have this risk when using a backup software. That's why I started to look for an alternative and found Restic, which seems (at this time) to work better and I'm on halfway of switching to it fully. Of course there is a possibility to create complex software, which still works perfectly, but the more features it has the more complex it gets and the more harder it gets to keep it running smoothly. Just my two cents, why I prefer it not to have any OS/file-system specific support out of Go-s own world.

**Comment by aventrax** [comment]
My slight modified version of  @turnkey-commerce  script
```
#
# Restic VSS Backup Script v0.1
#

# Environmental Variables
$env:RESTIC_PASSWORD = "...."
$env:RESTIC_REPOSITORY = "...."



# Paths to backup
$paths = @{}
$paths["C:\"] = @(
    'nginx',
    'test',
    'test with spaces'
)
#$paths["D:\"] = @(
#    'Software'
#)



# -- DO NOT MODIFY ABOVE -- #

ForEach ($item in $paths.GetEnumerator()) {

    # Create the Shadow Copy
    $s1 = (Get-WmiObject -List Win32_ShadowCopy).Create($item.Key, "ClientAccessible")
    $s2 = Get-WmiObject -Class Win32_ShadowCopy | Where-Object { $_.ID -eq $s1.ShadowID }

    $device  = $s2.DeviceObject + "\"
    $ShadowPath = $item.Key + 'VSS\'
    
    # Create a symbolic link to the shadow copy
    cmd /c mklink /d $ShadowPath "$device"

    # Build the new list of folders (incuding the ShadowPath)
    $folder_list = New-Object System.Collections.Generic.List[System.Object]
    ForEach ($path in $item.Value) {
        $p = $ShadowPath + $path
        $folder_list.Add($p)
    }

    # Print the command line
    cmd /c echo Launching: $resticExe -q backup $folder_list

    # Launch Restic
    cmd /c restic.exe -q backup $folder_list
    
    # Delete the shadow copy and remove the symbolic link
    $s2.Delete()
    cmd /c rmdir $ShadowPath
}
```

**Comment by jirib** [comment]
> Another issue that makes me wonder if it is a good idea to have VSS integrated into restic is: how would users configure it? From what you're describing it sounds like there's a high chance that we will run into a configuration nightmare...

For example burp[1] backup uses VSS by default on Windows and nothing has to be done (though the code seems to be originating from Bacula). They also provide a tool to strip VSS from backup files so one can restore files backuped on Windows onto Linux.

[1] https://github.com/grke/burp/search?q=vss&unscoped_q=vss

**Comment by bherila** [comment]
> They also provide a tool to strip VSS from backup files so one can restore files backuped on Windows onto Linux.

VSS just provides a consistent snapshot of the fs, it doesn't modify the files in any so I'm not sure what "stripping" does? Still agree that transparent VSS usage would be the friendliest way of using restic on Windows **in general** however I understand that it may not be the highest priority since it is a platform-specific feature. There are other things that (last I checked anyway) don't work too well in Windows, like FUSE mounting, which I would also love to see working!

**Comment by mholt** [comment]
After a lot of reading, then many hours of tinkering, and still not knowing fully what is going on, I was able to successfully create a shadow copy using, I think, pure Go, with the help of a couple libraries and some amazing luck.

Here's what I did:

Running cmd as Administrator:

```
C:\> vssadmin list shadows

vssadmin 1.1 - Volume Shadow Copy Service administrative command-line tool
(C) Copyright 2001-2013 Microsoft Corp.

No items found that satisfy the query.
```

Then I ran my Go program:

```
C:\> createshadowcopy.exe
```

Then again:

```
C:\> vssadmin list shadows
vssadmin 1.1 - Volume Shadow Copy Service administrative command-line tool
(C) Copyright 2001-2013 Microsoft Corp.

Contents of shadow copy set ID: {8786dacd-2567-466e-8d13-7711260f1c0e}
   Contained 1 shadow copies at creation time: 11/23/2018 1:30:35 PM
      Shadow Copy ID: {b9a647ff-9f8b-412e-a2f5-1b34d1ed9aad}
         Original Volume: (C:)\\?\Volume{6e7146fd-fb7a-494c-9421-346633d03dc5}\
         Shadow Copy Volume: \\?\GLOBALROOT\Device\HarddiskVolumeShadowCopy2
         Originating Machine: MYMACHINE
         Service Machine: MYMACHINE
         Provider: 'Microsoft Software Shadow Copy provider 1.0'
         Type: ClientAccessible
         Attributes: Persistent, Client-accessible, No auto release, No writers, Differential
```

It's worth noting that the Go program that did this was compiled from my Mac for Windows, then copied to a Windows VM, so it does still cross-compile.

The Go program must be run as Administrator, however.

Would you like to proceed?

I do have one thing to ask as a favor in return, though, to the Windows users in this thread: How are you using restic to back up mapped network drives, while being run as Administrator? Is there a better way to access drives than by their letters? (UNC paths?) And how do you present those to users? (i.e. Windows Explorer just shows drive letters I think...)

References:

- https://github.com/StackExchange/wmi/issues/6
- https://docs.microsoft.com/en-us/previous-versions/windows/desktop/vsswmi/create-method-in-class-win32-shadowcopy
- https://msdn.microsoft.com/en-us/library/windows/desktop/aa394428(v=vs.85).aspx

**Comment by turnkey-commerce** [comment]
That is very cool indeed! I would think it would be best if your tool could be imported as a package from restic so it can create and also presumably destroy the shadow copy of the local volume(s) that restic is backing up.

To answer your question, local volumes are usually referred to with drive letters and remote volumes are usually referred to with UNC paths.  Local shared volumes can be referred to either way on the local system if they are mapped to a letter drive.

So if you're trying to backup network drives you would normally use UNC.  Windows explorer is really using that when browsing the network and you can reveal what it is using by clicking in the path bar at the top to change it to a UNC path.

**Comment by bherila** [comment]
Since it's difficult to reliably VSS snapshot the remote machine, I typically recommend running restic on the file server if possible. This is ideal because the file server can be backed up even if other people are using the file share remotely.

**Comment by ProactiveServices** [comment]
@mholt Keep in mind that restic [cannot currently restore backups made from UNC paths](https://github.com/restic/restic/issues/2004#issuecomment-439078798). ISTR that when I tested a restore from a VSS path I had the same problem.

**Comment by mholt** [comment]
@turnkey-commerce 

> That is very cool indeed! I would think it would be best if your tool could be imported as a package from restic so it can create and also presumably destroy the shadow copy of the local volume(s) that restic is backing up.

Thanks. No need for a separate package though. If that PR happens in the linked issue over at the WMI package, it'll only be a few lines of code: just enough for its own function. If that PR doesn't happen, it's still just an extra function or two. Maybe its own code file. Or a fork of that package with the change.

> To answer your question, local volumes are usually referred to with drive letters and remote volumes are usually referred to with UNC paths. Local shared volumes can be referred to either way on the local system if they are mapped to a letter drive.

Interesting, so how does this work with the possibility that drive letters can change, and how does a program which is running as Administrator back up some user's C:\ or Z:\ or (whatever):\ drives?

@ProactiveServices 

> Keep in mind that restic cannot currently restore backups made from UNC paths. ISTR that when I tested a restore from a VSS path I had the same problem.

Alright, well, that can probably be fixed too.

**Comment by mholt** [comment]
@turnkey-commerce 

> and a program running as administrator would see the same drive letters as the user on the machine.

Hmm, this is not my experience. When running as administrator, I can't access all the drive letters the user has.

**Comment by turnkey-commerce** [comment]
@mholt 

> Hmm, this is not my experience. When running as administrator, I can't access all the drive letters the user has.

Sorry, you're right, I was thinking they are global but shows that I work on single-user machines too much.  Most of the times I have had to do something like that I create a UNC mapped drive using symbolic links (mklink /D...) which allows it to be accessible as both a standard and elevated user. However, that's not really drive letters so my bad. Drive letters are indeed a messy thing on Windows!

**Comment by bherila** [comment]
CIMv2 management objects have a standard method "Delete" ... does this help?

https://books.google.com/books?id=K1PzlJYEuz0C&pg=PA360&lpg=PA360&dq=Win32_ShadowCopy+delete+method&source=bl&ots=AJ87yYdsas&sig=m6LnGJdcGQaV2kmJw7oYWZ3bO5k&hl=en&sa=X&ved=2ahUKEwjbufK22PLeAhWIxVkKHfv0CFEQ6AEwCHoECAIQAQ#v=onepage&q=Win32_ShadowCopy%20delete%20method&f=false

**Comment by mholt** [comment]
Yeah, that'll probably do.

Probably relevant docs:

- https://docs.microsoft.com/en-us/windows/desktop/cimwin32prov/delete-method-in-class-win32-directory
- https://docs.microsoft.com/en-us/windows/desktop/WmiSdk/retrieving-a-class
- https://www.reddit.com/r/PowerShell/comments/3gza8n/removing_shadow_copies_with_cim_vs_wmi/

**Comment by bherila** [comment]
This is excellent; however, calling the API (as opposed to using wmic or vssadmin) has its advantages:
* Access to the return code HRESULT for handling errors
* Can take advantage of VSS auto releasing the snapshot when the calling process ends. See here for a list of VSS flags: https://docs.microsoft.com/en-us/windows/desktop/api/vss/ne-vss-_vss_volume_snapshot_attributes
*  No need for string parsing of stdout

**Comment by joonas-fi** [comment]
My package now transparently supports Linux (via LVM) also. I won't spam this issue more.

**Comment by nioncode** [comment]
In a way, I feel like restic should not be responsible to cover snapshots on all systems. On Windows, it just needs VSS, but on Linux it might use either LVM or file system specific snapshot technologies (e.g. on btrfs).

However, since there is a universal way for doing snapshots on Windows and this can be added relatively easy through command line invocations, I think it makes sense to include VSS support in restic. On Linux, it is quite trivial to create snapshots before running restic and deleting them afterwards (or keep them for a local history of files).

**Comment by cdhowie** [comment]
> I feel like restic should not be responsible to cover snapshots on all systems.

I think it makes sense to have restic support snapshots everywhere there is a consistent and unified service that will take care of the details.  On Windows, this is VSS.  On Linux, this doesn't exist yet, but it might some day when systemd decides it needs to do yet another thing.

**Comment by MB512** [comment]
I found this short looking documentation from MS to create a snapshot for backup. Maybe this could help someone.

https://docs.microsoft.com/en-us/windows/desktop/VSS/simple-shadow-copy-creation-for-backup

**Comment by mholt** [comment]
If someone wants to help wrap up my pull request over at https://github.com/StackExchange/wmi/pull/45 then we can probably make this possible in restic.

**Comment by fgma** [comment]
I've created a pull request to solve this a few weeks ago. Maybe someone is interested in testing it and give some feedback.

See https://github.com/restic/restic/pull/2274 for details.

**Comment by jcasale** [comment]
It appear there is significant oversight here. Without consideration for writers, the operator must have sufficient knowledge of Windows VSS (not likely in most cases) otherwise they will succumb to confusing errors.

Consider a test host with SQL Server installed and data spread across more than one drive:
```
C:\Users\Administrator\Desktop>sqlcmd -W -Q "SELECT name, physical_name FROM sys.master_files WHERE database_id = DB_ID('test')"
name physical_name
---- -------------
test c:\test.mdf
test_log E:\test_Log.ldf
```

Now lets list the writers, (output significantly truncated):
```
C:\>diskshadow
Microsoft DiskShadow version 1.0
Copyright (C) 2013 Microsoft Corporation
On computer:  IDM,  6/2/2019 2:36:27 PM


DISKSHADOW> LIST WRITERS

...
        * WRITER "SqlServerWriter"
                - Writer ID   = {a65faa63-5ea8-4ebc-9dbd-a0c4db26912a}
                - Writer instance ID = {92504559-51c7-45f5-abdd-6857da8794c9}
                - Supports restore events = TRUE
                - Writer restore conditions = VSS_WRE_ALWAYS
                - Restore method = VSS_RME_RESTORE_IF_CAN_REPLACE
                - Requires reboot after restore = FALSE
                - Excluded files:
                + Component "SqlServerWriter:\HOSTNAME\master"
                        - Name: master
                        - Logical path: HOSTNAME
                        - Full path: \HOSTNAME\master
                        - Caption:
                        - Type: VSS_CT_FILEGROUP [2]
                        - Is selectable: TRUE
                        - Is top level: TRUE
                        - Notify on backup complete: TRUE
                        - Paths affected by this component:
                                - C:\Program Files\Microsoft SQL Server\MSSQL13.MSSQLSERVER\MSSQL\DATA
                        - Volumes affected by this component:
                               - \\?\Volume{2af08413-0000-0000-0000-501f00000000}\ [C:\]
                        - Component Dependencies:
...
                + Component "SqlServerWriter:\HOSTNAME\test"
                        - Name: test
                        - Logical path: HOSTNAME
                        - Full path: \HOSTNAME\test
                        - Caption:
                        - Type: VSS_CT_FILEGROUP [2]
                        - Is selectable: TRUE
                        - Is top level: TRUE
                        - Notify on backup complete: TRUE
                        - Paths affected by this component:
                                - E:\Microsoft SQL Server\MSSQL13.MSSQLSERVER\MSSQL\DATA
                        - Volumes affected by this component:
                               - \\?\Volume{f1a7c8cf-0000-0000-0000-100000000000}\ [E:\]
                        - Component Dependencies:
```

Note the `Volumes affected by this component` lines, two volumes must be included in the snapshot for the writer to activate.

This is where the problem arises.

If you do not explicitly verify writers, they are silently excluded. Note the exclusion of the writer verification:
```
C:\Users\Administrator\Desktop>dir E:
 Volume in drive E is Test
 Volume Serial Number is 5226-8C3C

 Directory of E:\

06/02/2019  02:48 PM         8,388,608 test_log.ldf
               1 File(s)      8,388,608 bytes
               0 Dir(s)  10,680,373,248 bytes free

C:\Users\Administrator\Desktop>dir Z:
The system cannot find the path specified.

C:\Users\Administrator\Desktop>diskshadow /s backup.dsh
Microsoft DiskShadow version 1.0
Copyright (C) 2013 Microsoft Corporation
On computer:  HOSTNAME,  6/2/2019 2:55:20 PM

-> SET VERBOSE ON
-> SET CONTEXT PERSISTENT
-> SET METADATA C:\Temp\metadata.tmp
The existing file will be overwritten.
-> # WRITER VERIFY {a65faa63-5ea8-4ebc-9dbd-a0c4db26912a}
-> BEGIN BACKUP
-> ADD VOLUME e: ALIAS shadow_vol_e
-> CREATE
Excluding writer "Shadow Copy Optimization Writer", because all of its components have been excluded.
Component "\TasksStore" from writer "Task Scheduler Writer" is excluded from backup,
because it requires volume C:\ which is not in the shadow copy set.
...
Component "\HOSTNAME\master" from writer "SqlServerWriter" is excluded from backup,
because it requires volume C:\ which is not in the shadow copy set.
Component "\HOSTNAME\model" from writer "SqlServerWriter" is excluded from backup,
because it requires volume C:\ which is not in the shadow copy set.
Component "\HOSTNAME\msdb" from writer "SqlServerWriter" is excluded from backup,
because it requires volume C:\ which is not in the shadow copy set.
...
Component "\HOSTNAME\test" from writer "SqlServerWriter" is excluded from backup,
because it requires volume C:\ which is not in the shadow copy set.
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
Component "\BCD\BCD" from writer "ASR Writer" is excluded from backup,
because it requires volume  which is not in the shadow copy set.
...

Alias shadow_vol_e for shadow ID {b6998137-a589-4aef-ad5c-0f075522ebd3} set as environment variable.
Alias VSS_SHADOW_SET for shadow set ID {965642f7-1205-49ef-89e9-31dddbe41037} set as environment variable.
...

Querying all shadow copies with the shadow copy set ID {965642f7-1205-49ef-89e9-31dddbe41037}

        * Shadow copy ID = {b6998137-a589-4aef-ad5c-0f075522ebd3}               %shadow_vol_e%
                - Shadow copy set: {965642f7-1205-49ef-89e9-31dddbe41037}       %VSS_SHADOW_SET%
                - Original count of shadow copies = 1
                - Original volume name: \\?\Volume{f1a7c8cf-0000-0000-0000-100000000000}\ [E:\]
                - Creation time: 6/2/2019 2:55:40 PM
                - Shadow copy device name: \\?\GLOBALROOT\Device\HarddiskVolumeShadowCopy2
                - Originating machine: hostname
                - Service machine: hostname
                - Not exposed
                - Provider ID: {b5946137-7b9f-4925-af80-51abd60b20d5}
                - Attributes:  No_Auto_Release Persistent Differential

Number of shadow copies listed: 1
-> EXPOSE %shadow_vol_e% z:
-> %shadow_vol_e% = {b6998137-a589-4aef-ad5c-0f075522ebd3}
The shadow copy was successfully exposed as z:\.
-> END BACKUP
-> EXIT

C:\Users\Administrator\Desktop>dir Z:
 Volume in drive Z is Test
 Volume Serial Number is 5226-8C3C

 Directory of Z:\

06/02/2019  02:48 PM         8,388,608 test_log.ldf
               1 File(s)      8,388,608 bytes
               0 Dir(s)  10,684,567,552 bytes free
```

Now, with writer verification enabled:
```
C:\Users\Administrator\Desktop>diskshadow
Microsoft DiskShadow version 1.0
Copyright (C) 2013 Microsoft Corporation
On computer:  IDM,  6/2/2019 3:07:36 PM


DISKSHADOW> DELETE SHADOWS ALL
Deleting shadow copy {b6998137-a589-4aef-ad5c-0f075522ebd3} on volume \\?\Volume{f1a7c8cf-0000-0000-0000-100000000000}\ from provider {b5946137-7b9f-4925-af80-51abd60b20d5} [Attributes: 0x00120009]...

Number of shadow copies deleted: 1

DISKSHADOW>
C:\Users\Administrator\Desktop>diskshadow /s backup.dsh
Microsoft DiskShadow version 1.0
Copyright (C) 2013 Microsoft Corporation
On computer:  HOSTNAME,  6/2/2019 3:08:34 PM

-> SET VERBOSE ON
-> SET CONTEXT PERSISTENT
-> SET METADATA C:\Temp\metadata.tmp
The existing file will be overwritten.
-> WRITER VERIFY {a65faa63-5ea8-4ebc-9dbd-a0c4db26912a}
-> BEGIN BACKUP
-> ADD VOLUME e: ALIAS shadow_vol_e
-> CREATE
Excluding writer "Shadow Copy Optimization Writer", because all of its components have been excluded.
Component "\TasksStore" from writer "Task Scheduler Writer" is excluded from backup,
because it requires volume C:\ which is not in the shadow copy set.
...
Component "\HOSTNAME\master" from writer "SqlServerWriter" is excluded from backup,
because it requires volume C:\ which is not in the shadow copy set.
Component "\HOSTNAME\model" from writer "SqlServerWriter" is excluded from backup,
because it requires volume C:\ which is not in the shadow copy set.
Component "\HOSTNAME\msdb" from writer "SqlServerWriter" is excluded from backup,
because it requires volume C:\ which is not in the shadow copy set.
...
Component "\HOSTNAME\test" from writer "SqlServerWriter" is excluded from backup,
because it requires volume C:\ which is not in the shadow copy set.
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
...
The writer "WMI Writer" is now entirely excluded from the backup,
because it does not contain any components that can be included.

ERROR: The writer "{a65faa63-5ea8-4ebc-9dbd-a0c4db26912a}" was not included. Aborting ...
       Check the syntax of the writer name and writer ID.
```

**Comment by fgma** [comment]
@jcasale Thanks for the detailed analysis. To summarize it:

- enabling VSS writers will provide a consistent snapshot even when data is on different volumes
- each writer provides information about its affected paths

my proposed solution would be:

- enable all writers for a snapshot
- include all volumes in the snapshot that are affected by any writer that affects the current volume

To get a consistent backup of e.g. SQLServer one would need to include all affected paths manually in the backup as this depends on the data you want to backup.

**Comment by jcasale** [comment]
The problem is significantly complicated, take for example MS Exchange where if you include its associated writer and facilitate it by adding all related volumes, it truncates its logs when the backup completes.

That is a horrible side effect with a detrimental impact to an existing backup strategy if it relies on the logs remaining available.

I would recommend never automatically doing anything for the operator, but rather verbosely display exclusions for them. Provide a means to verify or exclude writers, roughly implement a subset of the useful commands `diskshadow` offers.

This provides the operator with no unexpected caveats but allows them to ensure their use case remains valid. For example, if I am actually backing up Exchange and I automate my script, by verifying the writer, my backup will fail if another admin adds a new volume later. This is consistent with a robust process free of surprises. I could then also explicitly exclude other writers I don't want involved and filter out paths on the volume that are unrelated to my use case.

By the way, thanks for work here and for the effort towards hashing out a great implementation.

**Comment by fbkopp** [comment]
In my opinion, Restic is not a tool to create backup as an operating system image.
I believe Writers are just to back up the system image as a whole, inside the operating system where the backup is being done. Restic has always been a file backup tool, folders, non-operating system image, restic can be used to backup virtual machine disks. Using writers does not make much sense in backing up the SQL Server folder with database data for example. For me, it makes more sense to back up files generated from a SQL Server backup, because the process of restoring a backup of SQL Server is much easier and more intuitive than restoring an installation of it.

**Comment by zoggdotorg** [comment]
I beg to differ fbkopp. A specific case is Outlook which locks the Outlook data files. So any use of a backup program to back up personal files on Windows really requires VSS support. Or at least that's what I find.

**Comment by iceymanz** [comment]
fbkopp I think you're confused. While VSS freezes a volume, it has nothing to do with operating system images. 

We need VSS support to access consistent copies of files locked by applications (of which there are many in a running system).  

VSS writers specific to SQL and Exchange are another step more complex as detailed in links below, and are probably unnecessary unless restic wants to specifically target the enterprise space.

A Guide for SQL Server Backup Application Vendors (dated but all I could find quickly):
https://docs.microsoft.com/en-us/previous-versions/sql/sql-server-2005/administrator/cc966520(v=technet.10)

An image showing all the steps for a SQL VSS quiece and backup:
https://docs.microsoft.com/en-us/previous-versions/sql/sql-server-2005/administrator/images/cc966520.sqlwriter03%28en-us%2ctechnet.10%29.gif

**Comment by fbkopp** [comment]
Maybe I was not very clear in the example of MS SQL Server. As shown by @iceymanz it is a very complicated process to back up. But not only backing up, you have to be able to restore, there are much more practical specific tools for this function. It has PR #2274 which covers file backup using VSS and can backup (locked) outlook files, thunderbird, among others for example.

**Comment by fgma** [comment]
My main idea for the pull-request was the basic use-case @iceymanz described: Backing up locked files e.g. outlook or thunderbird profiles.

For more complex enterprise scenarios I'm lacking the knowledge. I might upgrade the VSS integration to cover those cases as a second step if someone can help to create a concept how to do so.

**Comment by kmwoley** [comment]
Based on the help in this bug/thread, I created a whole set of scripts to automate Restic backup on Windows - including VSS support. Hope this helps someone else out there: https://github.com/kmwoley/restic-windows-backup

**Comment by rawtaz** [comment]
FWIW, I use @kmwoley's [restic-windows-backup](https://github.com/kmwoley/restic-windows-backup), and it works great. If one needs VSS with restic right now, just use that and be happy :)

**Comment by azurefreecovid** [comment]
I just found @fgma's [PR](https://github.com/restic/restic/pull/2274) and have started using it as it is by far the best way I've seen to use Restic with VSS. Without VSS on Windows, any backup utility, including Restic is significantly reduced in its usefulness.

Anyway I've been testing @fgma's PR all morning and it seems to be working flawlessly (for my use case's anyway). I'm so comfortable with it I've switched all my Windows machines over to this branch. 

If other people out there are looking for a Windows VSS solution for Restic I'd highly encourage you to give the PR a try. If more people try it and test it I should imagine there is a higher change it will be included in the Restic code base.

Given Restic is all written in Go, it is trivial to compile a binary. I cross compiled mine on my Linux box and it takes no time at all. If you are new to compiling Go there are some great instructions in the PR itself at this [comment](https://github.com/restic/restic/pull/2274#issuecomment-627233355) which will point you in the right direction.

**Comment by rawtaz** [comment]
There's an updated restic binary that includes the VSS support in @fgma's PR at: https://beta.restic.net/pr-2274/ - please try it and report any problems!
